rust/bindgen.bzl

```diff
@@ -0,0 +1,156 @@
+# Copyright 2015 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+def _rust_bindgen(ctx):
+    rust_toolchain = ctx.toolchains["@io_bazel_rules_rust//rust:toolchain"]
+
+    bindgen = ctx.executable.bindgen
+    rustfmt = rust_toolchain.rustfmt if hasattr(rust_toolchain, 'rustfmt') else None
+    clang = ctx.executable.clang
+    libclang = ctx.attr.libclang
+    libstdcxx = ctx.attr.libstdcxx
+    cc_toolchain = ctx.attr.cc_toolchain
+    cc_fragment = ctx.fragments.cpp
+
+    # nb. We can't grab the cc_library`s direct headers, so a header must be provided.
+    cc_lib = ctx.attr.cc_lib
+    if not hasattr(cc_lib, "cc"):
+        fail("{} is not a cc_library".format(cc_lib))
+    header = ctx.file.header
+    if header not in cc_lib.cc.transitive_headers:
+        fail("Header {} is not in {}'s transitive closure of headers.".format(ctx.attr.header, cc_lib))
+
+    # rustfmt is not in the usual place, so bindgen would fail to find it
+    bindgen_args = ["--no-rustfmt-bindings"]
+    clang_args = []
+
+    output = ctx.outputs.out
+
+    clang_libs = depset(libclang.cc.libs + libstdcxx.cc.libs)
+    include_directories = depset(
+        cc_fragment.built_in_include_directories + [f.dirname for f in cc_lib.cc.transitive_headers]
+    )
+
+    if rustfmt:
+        unformatted = ctx.actions.declare_file(output.basename + ".unformatted")
+    else:
+        unformatted = output
+
+    args = ctx.actions.args()
+    args.add_all(bindgen_args)
+    args.add(header.path)
+    args.add("--output", unformatted.path)
+    args.add("--")
+    args.add_all(include_directories, before_each="-I")
+    args.add_all(clang_args)
+    ctx.actions.run(
+        inputs=depset(
+            [header],
+            # Standard library headers aren't in the transitive_headers, so we find them in the cc_toolchain
+            transitive=[cc_lib.cc.transitive_headers, clang_libs, cc_toolchain.files],
+        ),
+        outputs=[unformatted],
+        mnemonic="RustBindgen",
+        progress_message="Generating bindings for {}..".format(header.path),
+        env={
+            "RUST_BACKTRACE": "1",
+            # Bindgen loads libclang at runtime, so we setup LD_LIBRARY_PATH
+            "LD_LIBRARY_PATH": ":".join([f.dirname for f in clang_libs]),
+            "CLANG_PATH": clang.path,
+        },
+        executable=bindgen,
+        arguments=[args],
+        tools=[bindgen, clang],
+    )
+
+    if rustfmt:
```

Comment:
I think it's probably a good idea to separate the presence of the rustfmt binary from the decision to use rustfmt. Does it make sense to add a "should_fmt" bool?

Reply:
If including rustfmt is an optional part of the toolchain configuration, I don't see why one would go through the trouble of including it and not want it to apply to the generated code. But adding the option is fine

Reply:
Several points:

- I'd pursue a workflow where this command was exposed as an alias with the binaries pre-populated. That means anyone working in my workspace has the tool ready to use. The current implementation can't support this approach as providing the binary would prevent callers from disabling formatting
- All would-be formatters need to know the path to the rustfmt binary. In a large repo this doesn't seem like a reasonable expectation. This applies to some of the other args here (clang/libclang) hence point 1.

Reply:
I'm not sure if by "alias" you mean the hypothetical simplified local version of the rule or the binary itself. 

**Former:**

I'd follow my ongoing conventions and put this under //tools/rust:rust_bindgen

**Latter:**

If rustfmt comes from a rule parallel to the toolchain per https://github.com/bazelbuild/rules_rust/issues/101, i'd expect it to be brought in via a repository_rule like
```python
rust_extended_tools(
  name = "any_name",
  rustfmt_version = "0.6.1"
)
```
and would thus live under "@any_name//:rustfmt" or "@any_name//:bin/rustfmt".

If not that, then I'd expect this to be brought in by cargo. I prefer following Google's internal policies for internal deps and bring them into the working tree, so I'd expect it to be called something like "//third_party/cargo:cargo_bin_rustfmt".

Reply:
By 'alias' I meant the curried version of the rule, yes.

iirc the reason to avoid //third_party is because it doesn't work with recursive workspaces as well -- definitely punting on canonized naming for now.

Reply:
I think this is also resolved by using a `toolchain` instead.

Reply:
Yeah this SGTM